oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/DefaultOAuth2AuthorizationGrant.java

```diff
@@ -100,54 +94,52 @@ class DefaultOAuth2AuthorizationGrant implements OAuth2AuthorizationGrant {
-            if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
-                break;
-            }
+    private AsyncLoader<GrantedOAuth2AccessToken> tokenLoader() {
+        if (tokenLoader != null) {
+            return tokenLoader;
         }
 
-        final CompletableFuture<GrantedOAuth2AccessToken> newTokenFuture = future;
-        if (token == null && fallbackTokenProvider != null) {
-            CompletableFuture<? extends GrantedOAuth2AccessToken> fallbackTokenFuture = null;
-            try {
-                fallbackTokenFuture = requireNonNull(
-                        fallbackTokenProvider.get(), "fallbackTokenProvider.get() returned null");
-            } catch (Exception e) {
-                logger.warn("Unexpected exception from fallbackTokenProvider.get()", e);
-            }
-            if (fallbackTokenFuture != null) {
-                fallbackTokenFuture.handle((storedToken, unused) -> {
-                    if (isValidToken(storedToken)) {
-                        newTokenFuture.complete(storedToken);
+        final Function<GrantedOAuth2AccessToken, CompletableFuture<GrantedOAuth2AccessToken>> loader =
+                token -> {
+            final CompletableFuture<GrantedOAuth2AccessToken> newTokenFuture = new CompletableFuture<>();
+
+            if (token == null && fallbackTokenProvider != null) {
+                CompletableFuture<? extends GrantedOAuth2AccessToken> fallbackTokenFuture = null;
+                try {
+                    fallbackTokenFuture = requireNonNull(
+                            fallbackTokenProvider.get(), "fallbackTokenProvider.get() returned null");
+                } catch (Exception e) {
+                    logger.warn("Unexpected exception from fallbackTokenProvider.get()", e);
+                }
+                if (fallbackTokenFuture != null) {
+                    fallbackTokenFuture.handle((storedToken, unused) -> {
+                        if (isValidToken(storedToken)) {
+                            newTokenFuture.complete(storedToken);
+                            return null;
+                        }
+                        obtainAccessToken(newTokenFuture);
                         return null;
-                    }
-                    obtainAccessToken(newTokenFuture);
-                    return null;
-                });
+                    });
+                    return newTokenFuture;
+                }
+            }
+
+            if (token != null && token.isRefreshable()) {
+                refreshAccessToken(token, newTokenFuture);
                 return newTokenFuture;
             }
-        }
-        if (token != null && token.isRefreshable()) {
-            refreshAccessToken(token, newTokenFuture);
+
+            obtainAccessToken(newTokenFuture);
             return newTokenFuture;
-        }
-        obtainAccessToken(newTokenFuture);
-        return newTokenFuture;
+        };
+
+        tokenLoader = AsyncLoader.builder(loader)
```

Comment:
I don't think this is thread-safe. Many `tokenLoader` could be created in a multi-threading environment. Should we create `tokenLoader` in the constructor of `DefaultOAuth2AuthorizationGrant`?

Reply:
✅ updated